src/lsm/cache_map.zig

```diff
@@ -0,0 +1,358 @@
+                    gop.key_ptr.* = value.*;
+                }
+            }
+        }
+
+        pub fn remove(self: *Self, key: Key) void {
+            // The only thing that tests this in any depth is the cache_map fuzz itself.
+            // Make sure we aren't being called in regular code without another once over.
+            assert(constants.verify);
+
+            const maybe_removed = self.cache.remove(key);
+
+            if (maybe_removed) |removed| {
+                if (self.scope_is_active) {
+                    _ = self.scope_map.putAssumeCapacity(removed, {});
+                }
+            } else {
+                if (self.scope_is_active) {
+                    // TODO: Actually, does the fuzz catch this...
+                    // TODO: So if we delete from map_2 and put to map_1, there's a problem;
+                    //       because when we undo our scope we insert back to map_1 :/
+                    var maybe_map_removed = self.map_1.getKey(tombstone_from_key(key)) orelse
+                        self.map_2.getKey(tombstone_from_key(key));
+                    if (maybe_map_removed) |map_removed| {
+                        _ = self.scope_map.putAssumeCapacity(map_removed, {});
+                    }
+                }
+            }
+
+            // We always need to try remove from the stash; since it could have a stale value.
+            _ = self.map_1.remove(tombstone_from_key(key));
+            _ = self.map_2.remove(tombstone_from_key(key));
+        }
+
+        /// Start a new scope. Within a scope, changes can be persisted
+        /// or discarded. Only one scope can be active at a time.
+        pub fn scope_open(self: *Self) void {
+            assert(!self.scope_is_active);
+            self.scope_is_active = true;
+        }
+
+        pub fn scope_close(self: *Self, data: ScopeCloseMode) void {
+            assert(self.scope_is_active);
+            self.scope_is_active = false;
+
+            // We don't need to do anything to persist a scope.
+            if (data == .persist) {
+                self.scope_map.clearRetainingCapacity();
+                return;
+            }
+
+            // The scope_map stores the operations we need to reverse the changes a scope made.
+            // Replay them back.
+            var it = self.scope_map.iterator();
+            while (it.next()) |kv| {
+                const value = kv.key_ptr;
+                if (tombstone(value)) {
+                    // Reverting an insert consists of a .remove call. The value in here will be a
+                    // tombstone indicating the original value didn't exist. We don't touch map_2;
+                    // since we can never insert into it directly (only a .compact() can).
+                    const key = key_from_value(value);
+                    _ = self.cache.remove(key);
+                    _ = self.map_1.remove(tombstone_from_key(key));
```

Comment:
I don't believe we can (at least with the current map 'log' for scopes which stores the original value). For example, if the following sequence of events happens:
* `scope_open()`
* `upsert(123)` -> a tombstone for `123` is in the `scope_map`.
* `delete(123)` -> there is already an entry for `123` in the `scope_map` so nothing is changed.
* `scope_close(.discard)`

It'll try to remove `123`, but it will no longer exist in any layer due to the `delete`. I'll revisit this though after looking at switching the scope_map to be a log!

Reply:
(for the record, I didn't come up with this, the fuzzer found it when I added the assert :smile:)

Reply:
With it now being a log, the assert is back and passes the fuzz.